sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -464,6 +464,25 @@ class Dataset[T] private[sql](
    */
   def as[U : Encoder]: Dataset[U] = Dataset[U](sparkSession, logicalPlan)
 
+  /**
+   * Returns a new DataFrame where each row is reconciled to match the specified schema. Spark will:
+   * - Reorder columns and/or inner fields by name to match the specified schema.
+   * - Project away columns and/or inner fields that are not needed by the specified schema. Missing
+   *   columns and/or inner fields lead to failures.
+   * - Cast the columns and/or inner fields to match the data types in the specified schema, if the
+   *   types are compatible, e.g., numeric to numeric (error if overflows), but not string to int.
+   * - The columns and/or inner fields will carry the metadata from the specified schema, while
+   *   still keep their own metadata if not overwritten by the specified schema.
+   * - Fail if the nullability are not compatible. For example, the column and/or inner field is
+   *   nullable but the specified schema requires them to be not nullable.
```

Comment:
Are we talking about this? This sounds same to me.
Do we fail if the column is non-nullable but the specified schema is nullable? That should work I think. e.g.)

```scala
scala> spark.range(1).toDF.union(spark.createDataFrame(Tuple1(None) :: Tuple1(Some(1)) :: Nil)).printSchema()
root
 |-- id: long (nullable = true)


scala> spark.range(1).toDF.printSchema()
root
 |-- id: long (nullable = false)


scala> spark.createDataFrame(Tuple1(None) :: Tuple1(Some(1)) :: Nil).printSchema()
root
 |-- _1: integer (nullable = true)
```

Reply:
I don't have a strong opinion. My worry is `the column is non-nullable but the specified schema is nullable` could be a common case, as people tend to be conservative and always use nullable.

Reply:
My question is: what can go wrong if you expect a nullable column but it's actually non-nullable? If no I prefer the current behavior.

Reply:
Alright. I agree that 1. the implementation would require some expressions that currently do not exist in Spark SQL, 2.  the current implementation addresses most cases 3. nullability case is rather a corner case, and doesn't break how Spark SQL works in general.

Since we already have inconsistency about nullability (e.g., `DataFrameReader.schema`, etc.), I am good with leaving this as a future work.

Reply:
Maybe let's add Experimental tag though ..